### PR TITLE
Add GitHub Actions workflow for publishing pre-release to npm

### PR DESCRIPTION
**Pull request by syedsalehinipg**
## Changes

This PR introduces a new GitHub Actions workflow designed to automate the publishing of pre-release versions of the React-UI library to npm whenever changes are pushed to release/v* branches.

#### Steps - The workflow is triggered by pushes to `release/v*` branches
      
1. **Checkout Code:** Uses `actions/checkout@v2` to fetch the code from the repository with full history (`fetch-depth: 0`).      

2. **Fetch Main Branch:** Fetches the main branch from the remote repository.
      
3. **Check for Merge Commits:**
    - Finds the commit where the current branch diverged from the main branch.
    - Logs the merge commits since that point, excluding certain patterns.
    - Saves merge commit messages in `commit_messages.txt`.
    - Counts the merge commits and sets outputs based on the count.

      
4. **Cancel Workflow if No Merge Commits:** Uses `andymckay/cancel-action@0.2` to cancel the workflow if no merge commits are found.
      
5. **Set Up Node.js:** Configures the environment with Node.js 20.x using `actions/setup-node@v2` and sets the npm registry URL.
      
6. **Install pnpm:** Installs pnpm globally using npm.
      
7. **Install Dependencies:** Runs `pnpm install` to ensure all necessary dependencies are installed.
      
8. **Run Tests:**
    - Executes `pnpm test` to verify that all unit tests pass before proceeding.
    - Executes `pnpm test:docker` to run additional tests.
      
9. **Bump Version:**
    - Uses `pnpm version prerelease --no-git-tag-version` to update the package version for pre-releases without creating a git tag.
    - Configures git user details for the action.
    - Sets the new version in the GitHub environment variable.
      
10. **Configure npm for Scoped Packages:** Adds npm authentication details to `.npmrc` for scoped packages.
      
11. **Publish to npm:** Publishes the package to npm under the `next` tag using the NPM_TOKEN secret for authentication.
      
12. **Generate Release Notes from Merge Commit Messages:** Generates release notes based on the merge commits and saves them to `merge_commits.txt`.
      
13. **Commit Version Changes:**
    - Commits the version changes to the repository.
    - Creates a Git tag with the new version.
      
14. **Push Changes and Tags:** Pushes the changes and the newly created tag to the repository.
      
15. **Create GitHub Release:** Uses `softprops/action-gh-release@v1` to create a GitHub release with the tag, release notes, and marks it as a pre-release.
      
16. **Post Release Message to Teams:** Sends a message to a Microsoft Teams channel with details about the new pre-release version using a webhook URL.

**Authentication and Secrets:**

  - The workflow uses the `NPM_TOKEN` secret for npm authentication.
  - The `TEAMS_WEBHOOK_URL` secret is used to post a message to Microsoft Teams.

## Testing notes

1. Make a Small Commit to a Release Branch:

   -  Choose or create a release branch that follows your naming convention, such as release/v7.6.0.
   -  Make a minor but identifiable change, like updating a comment, documentation, or a trivial code adjustment that doesn’t affect functionality in a separate branch
     -   Merge that branch into the release branch, ensure the commit message clearly indicates the change you want to show in the release notes. e.g. Added new component call ABC

3. Ensure the Action Runs and Completes Successfully:

   -  After pushing your commit, navigate to the Actions tab in your GitHub repository.
   -  Select the workflow run triggered by your push to the release branch.
   -  Monitor the workflow to ensure each step completes without errors. Pay special attention to:
        -  The Install Dependencies step to confirm all necessary packages are installed correctly.
        -  The Run Tests step to ensure all tests pass without any issues.
        -  The Bump Version and Create Tag step to verify the version is incremented appropriately.
        -  The Publish to npm step for successful execution and confirmation that the package is pushed to npm.
        -  The release notes are published and a message is sent to the React-UI Teams channel. 

4. Check npm Registry for the Correct Pre-release Version:
   -  Go to the npm registry (or your private npm registry if used) and locate your package.
   -  Confirm that the new pre-release version (e.g., 7.6.0-0) is published.
   -  Ensure that the version is tagged correctly with next, indicating it's a pre-release and not the latest stable version.

## Author checklist before assigning a reviewer

- [x] Reviewed my own code-diff.
- [ ] Branch has been run in docker.
- [x] PR assigned to me or an appropriate delegate.
- [x] Relevant labels added to the PR.
- [ ] Appropriate tests have been added.
- [x] Lint and test workflows pass.

### CODE REVIEW

**Comment by dmbarry86** [comment]
Are you planning to also automate the release on GitHub and the release notes?

**Comment by syedsalehinipg** [comment]
> Are you planning to also automate the release on GitHub and the release notes?

I am, I was going do this separately as I have to look into how to do. But I can do it part of this PR no problem.

**Comment by dmbarry86** [comment]
> I am, I was going do this separately as I have to look into how to do. But I can do it part of this PR no problem.

Makes sense to do together I think. Can't really have one without the other IMO.

File: .github/workflows/npm-prerelease.yml [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        uses: actions/setup-node@v2
+        with:
+          node-version: "20.x"
+          registry-url: "https://registry.npmjs.org/"
+
+      - name: Install Dependencies
+        run: pnpm install
+
+      - name: Run Tests
+        run: pnpm test
```
Comment (dmbarry86) [inline .github/workflows/npm-prerelease.yml:25]: If this step fails will it stop the release or will the step afterwards still run? I don't think we want to do a release if the tests fail.

Comment (dmbarry86) [reply .github/workflows/npm-prerelease.yml:25]: We should probably also run these tests via Docker, like we do for the test action.

Comment (syedsalehinipg) [reply .github/workflows/npm-prerelease.yml:25]: > If this step fails will it stop the release or will the step afterwards still run? I don't think we want to do a release if the tests fail.

 If the step fails, the workflow stops here and does not proceed to the following steps.

**Comment by syedsalehinipg** [comment]
GitHub Actions workflow should do the following when a push is made to a branch named like `release/v*`:

1. Checkout the code.
2. Set up Node.js.
3. Install dependencies.
4. Run tests.
5. Bump the version and create a git tag.
6. Generate release notes from commit messages.
7. Push the tags to the repository.
8. Publish the package to npm.
9. Create a GitHub release.
10. Post a release message to Microsoft Teams.

**Comment by dmbarry86** [comment]
@syedsalehinipg what release number do you suggest using when I branch to test out this workflow? What release numbers are currently planned?

**Comment by syedsalehinipg** [comment]
> @syedsalehinipg what release number do you suggest using when I branch to test out this workflow? What release numbers are currently planned?

I think the latest release available.

Currently planned:

- v6.3.1
- v7.7.0
- v8.0.0

**Comment by dmbarry86** [comment]
> I think the latest release available.
> 
> Currently planned:
> 
> * v6.3.1
> * v7.7.0
> * v8.0.0

Just to be clear (so I don't mess up your releasing), you are suggesting to use the latest so version 8 and do a pre-release against that?

**Comment by syedsalehinipg** [comment]
> > I think the latest release available.
> > Currently planned:
> > 
> > * v6.3.1
> > * v7.7.0
> > * v8.0.0
> 
> Just to be clear (so I don't mess up your releasing), you are suggesting to use the latest so version 8 and do a pre-release against that?

That's correct yes.

**Comment by dmbarry86** [comment]
Have you tried this out yourself yet? It seems to fail at the first step during install for me.

https://github.com/IPG-Automotive-UK/react-ui/actions/runs/9091917772/job/24987485440

I created branch release/v8.0.0-2 from release/v8.0.0 and then merged in this branch related to this PR.

**Review by dmbarry86 — CHANGES_REQUESTED** [review]
Fails at first step. I can't really test any further.

**Comment by syedsalehinipg** [comment]
The PR details have been updated to reflect the changes made.

**Comment by dmbarry86** [comment]
Just tried an example but it seems to fail at the npm publish step. Any thoughts?

https://github.com/IPG-Automotive-UK/react-ui/actions/runs/9158522344/job/25177060257

**Review by dmbarry86 — CHANGES_REQUESTED** [review]
Marking this as changes required until the action issues are debugged and resolved so it appears off my list. Re-request a review from me when ready.

**Pull request closed without merging**